### PR TITLE
Inital implementation of KlasaConsole and Colors

### PR DESCRIPTION
**Pull request by MrJacz**
This PR implements KlasaConsole and Colors.
I currently am unsure of how to do the unit testing for the most part and would like someone else to take a crack at that if someone could!

### CODE REVIEW

File: src/lib/Colors.ts [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+	 * The opening tags
+	 */
+	public opening: string;
+
+	/**
+	 * the closing tags
+	 */
+	public closing: string;
+
+	constructor(options: ColorsFormatOptions = {}) {
```
Comment (bdistin) [inline src/lib/Colors.ts:60]:
```suggestion
	public constructor(options: ColorsFormatOptions = {}) {
```

File: src/lib/Colors.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+		this.opening = (this.constructor as typeof Colors).useColors && opening ? `\u001B[${opening.join(';')}m` : '';
+		this.closing = (this.constructor as typeof Colors).useColors && closing ? `\u001B[${closing.join(';')}m` : '';
+	}
+
+	/**
+	 * Format a string
+	 * @param string The string to format
+	 */
+	format(string: string): string {
```
Comment (bdistin) [inline src/lib/Colors.ts:71]:
```suggestion
	public format(string: string): string {
```

File: src/lib/Colors.ts [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+		return this.opening + string + this.closing;
+	}
+
+	/**
+		* Applies the style
+		* @param styles The style or styles to apply
+		* @param FormatData the format data
+		*/
+	static style(styles?: Style | Style[], { opening = [], closing = [] }: ColorsFormatData = {}): ColorsFormatData {
```
Comment (bdistin) [inline src/lib/Colors.ts:80]:
```suggestion
	private static style(styles?: Style | Style[], { opening = [], closing = [] }: ColorsFormatData = {}): ColorsFormatData {
```

File: src/lib/Colors.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+	 */
+	format(string: string): string {
+		return this.opening + string + this.closing;
+	}
+
+	/**
+		* Applies the style
+		* @param styles The style or styles to apply
+		* @param FormatData the format data
+		*/
```
Comment (bdistin) [inline src/lib/Colors.ts:79]:
```suggestion
	* Applies the style
	* @param styles The style or styles to apply
	* @param FormatData the format data
	*/
```

File: src/lib/Colors.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+			opening.push(this.TEXTS[text.toLowerCase()]);
+			closing.push(this.CLOSE.text);
+		}
+		return { opening, closing };
+	}
+
+	/**
+	 * Determines if this class should be constructed with colors or not
+	 */
+	static useColors = true;
```
Comment (bdistin) [inline src/lib/Colors.ts:129]:
```suggestion
	protected static useColors = true;
```

File: src/lib/KlasaConsole.ts [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+	 */
+	shard: ColorsFormatOptions;
+}
+
+/**
+ * Klasa's console class, extends NodeJS Console class.
+ */
+export class KlasaConsole extends Console {
+
+
```
Comment (bdistin) [inline src/lib/KlasaConsole.ts:86]:
```suggestion

```

File: src/lib/Colors.ts [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+/**
+ * The Colors class that manages the colors displayed in the console.
+ */
+export class Colors {
+
+	/**
+	 * The opening tags
+	 */
+	public opening: string;
```
Comment (bdistin) [inline src/lib/Colors.ts:53]:
```suggestion
	public readonly opening: string;
```

File: src/lib/Colors.ts [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+	/**
+	 * the closing tags
+	 */
+	public closing: string;
```
Comment (bdistin) [inline src/lib/Colors.ts:58]:
```suggestion
	public readonly closing: string;
```

File: src/lib/KlasaConsole.ts [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+	 * Whether the timestamp should be in utc or not
+	 */
+	public utc: boolean;
+
+	/**
+	 * The standard output stream for this console, defaulted to process.stderr.
+	 */
+	private stdout: NodeJS.WriteStream;
+
+	constructor(options: Partial<ConsoleOptions> = {}) {
```
Comment (bdistin) [inline src/lib/KlasaConsole.ts:107]:
```suggestion
	public constructor(options: Partial<ConsoleOptions> = {}) {
```

File: src/lib/KlasaConsole.ts [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+		const { time, message } = this.colors[type];
+		const timestamp = this.template ? time.format(`[${this.timestamp}]`) : '';
+		super[ConsoleTypes[type] || 'log'](content.split('\n').map(str => `${timestamp} ${message.format(str)}`).join('\n'));
+	}
+
+	/**
+	 * Calls a log write with everything to the console/writable stream.
+	 * @param data The data we want to print
+	 */
+	log(...data: any[]): void {
```
Comment (bdistin) [inline src/lib/KlasaConsole.ts:151]:
```suggestion
	public log(...data: any[]): void {
```

File: src/lib/KlasaConsole.ts [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+		this.write(data, 'log');
+	}
+
+	/**
+	 * Calls a warn write with everything to the console/writable stream.
+	 * @param data The data we want to print
+	 */
+	warn(...data: any[]): void {
```
Comment (bdistin) [inline src/lib/KlasaConsole.ts:159]:
```suggestion
	public warn(...data: any[]): void {
```

File: src/lib/KlasaConsole.ts [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+		this.write(data, 'warn');
+	}
+
+	/**
+	 * Calls an error write with everything to the console/writable stream.
+	 * @param data The data we want to print
+	 */
+	error(...data: any[]): void {
```
Comment (bdistin) [inline src/lib/KlasaConsole.ts:167]:
```suggestion
	public error(...data: any[]): void {
```

File: src/lib/KlasaConsole.ts [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+		this.write(data, 'error');
+	}
+
+	/**
+	 * Calls an error write with everything to the console/writable stream.
+	 * @param data The data we want to print
+	 */
+	debug(...data: any[]): void {
```
Comment (bdistin) [inline src/lib/KlasaConsole.ts:175]:
```suggestion
	public debug(...data: any[]): void {
```

File: src/lib/KlasaConsole.ts [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+		this.write(data, 'debug');
+	}
+
+	/**
+	 * Calls a verbose write with everything to the console/writable stream.
+	 * @param data The data we want to print
+	 */
+	verbose(...data: any[]): void {
```
Comment (bdistin) [inline src/lib/KlasaConsole.ts:183]:
```suggestion
	public verbose(...data: any[]): void {
```

File: src/lib/KlasaConsole.ts [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+		this.write(data, 'verbose');
+	}
+
+	/**
+	 * Calls a wtf (what a terrible failure) write with everything to the console/writable stream.
+	 * @param data The data we want to print
+	 */
+	wtf(...data: any[]): void {
```
Comment (bdistin) [inline src/lib/KlasaConsole.ts:191]:
```suggestion
	public wtf(...data: any[]): void {
```

File: test/Util.ts [hunk 16]
```diff
@@ -1,6 +1,6 @@
-import ava from 'ava';
-import util from '../src';
+// import ava from 'ava';
+// import util from '../src';
```
Comment (bdistin) [inline test/Util.ts:2]:
```suggestion
import ava from 'ava';
```

File: test/Util.ts [hunk 17]
```diff
@@ -1,6 +1,6 @@
...
 
-ava('sample test', (test): void => {
-	test.is(util(), 'bar');
-});
+// ava('sample test', (test): void => {
+// 	test.is(util(), 'bar');
+// });
```
Comment (bdistin) [inline test/Util.ts:6]:
```suggestion
ava('sample test', (test): void => {
 	test.pass();
});
```

File: src/lib/KlasaConsole.ts [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+
+		for (const [name, formats] of Object.entries(options.colors)) {
+			this.colors[name] = {};
+			for (const [type, format] of Object.entries(formats)) this.colors[name][type] = new Colors(format);
+		}
+
+		this.utc = options.utc;
+	}
+
+
```
Comment (bdistin) [inline src/lib/KlasaConsole.ts:124]:
```suggestion

```

File: src/lib/KlasaConsole.ts [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+	private stdout: NodeJS.WriteStream;
+
+	public constructor(options: Partial<ConsoleOptions> = {}) {
+		options = mergeDefault(ConsoleDefaults, options) as Required<ConsoleOptions>;
+		super(options.stdout, options.stderr);
+
+		Object.defineProperty(this, 'stdout', { value: options.stdout });
+		Object.defineProperty(this, 'stderr', { value: options.stderr });
+
+		Colors.useColors = typeof options.useColor === 'undefined' ? this.stdout.isTTY || false : options.useColor;
```
Comment (bdistin) [inline src/lib/KlasaConsole.ts:112]:
```suggestion

		// eslint-disable-next-line dot-notation
		Colors['useColors'] = typeof options.useColor === 'undefined' ? this.stdout.isTTY || false : options.useColor;
```

File: src/lib/KlasaConsole.ts [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+	 * Flattens data into a readable string.
+	 * @param data Data to flatten, could be anything
+	 */
+	private static _flatten(data: unknown | unknown[]): string {
+		if (typeof data === 'undefined' || typeof data === 'number' || data === null) return String(data);
+		if (typeof data === 'string') return data;
+		if (typeof data === 'object') {
+			const isArray = Array.isArray(data);
+			if (isArray && (data as []).every(datum => typeof datum === 'string')) return (data as string[]).join('\n');
+			return (data as any).stack || (data as any).message || inspect(data, { depth: Number(isArray), colors: Colors.useColors });
```
Comment (bdistin) [inline src/lib/KlasaConsole.ts:201]:
```suggestion
			if (isArray && (data as []).every(datum => typeof datum === 'string')) return (data as string[]).join('\n');
			// eslint-disable-next-line dot-notation
			return (data as any).stack || (data as any).message || inspect(data, { depth: Number(isArray), colors: Colors['useColors'] });
```

File: src/lib/KlasaConsole.ts [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+	 * A shard object containing colors and styles
+	 */
+	shard: ColorsFormatOptions;
+}
+
+/**
+ * Klasa's console class, extends NodeJS Console class.
+ */
+export class KlasaConsole extends Console {
+	/**
```
Comment (bdistin) [inline src/lib/KlasaConsole.ts:85]:
```suggestion
export class KlasaConsole extends Console {

	/**
```

File: test/Util.ts [hunk 22]
```diff
@@ -1,6 +1,5 @@
 import ava from 'ava';
-import util from '../src';
 
 ava('sample test', (test): void => {
-	test.is(util(), 'bar');
+ 	test.pass();
```
Comment (bdistin) [inline test/Util.ts:4]:
```suggestion
	test.pass();
```

File: src/lib/Colors.ts [hunk 23]
```diff
@@ -0,0 +1,233 @@
...
+	/**
+	 * @since 0.0.1
+	 * @param options The options.
+	 * @example
+	 * const red = new Colors({ text: 'red' });
+	 * console.log(red.format('Hello World'));
+	 * // A 'Hello World' string will be printed red.
+	 */
+	public constructor(options: ColorsFormatOptions = {}) {
+		const { opening, closing } = Colors.text(options.text, (this.constructor as typeof Colors).background(options.background, (this.constructor as typeof Colors).style(options.style)));
```
Comment (vladfrangu) [inline src/lib/Colors.ts:83]: Why is there inconsistency between `Colors.text` and `this.constructor`?

File: src/lib/Colors.ts [hunk 24]
```diff
@@ -0,0 +1,233 @@
...
+	}
+
+	/**
+	 * Applies the background.
+	 * @since 0.0.1
+	 * @param background the background to apply
+	 * @param FormatData the format data
+	 */
+	private static background(background?: Color, { opening = [], closing = [] }: ColorsFormatData = {}): ColorsFormatData {
+		if (background && background.toLowerCase() in this.BACKGROUNDS) {
```
Comment (vladfrangu) [inline src/lib/Colors.ts:124]: Let's use `Reflect.has` like everywhere else
```suggestion
		if (background && Reflect.has(this.BACKGROUNDS, background.toLowerCase())) {
```

File: src/lib/Colors.ts [hunk 25]
```diff
@@ -0,0 +1,233 @@
...
+	}
+
+	/**
+	 * Applies the text format.
+	 * @since 0.0.1
+	 * @param text the text format to apply
+	 * @param FormatData the format data
+	 */
+	private static text(text?: Color, { opening = [], closing = [] }: ColorsFormatData = {}): ColorsFormatData {
+		if (text && text.toLowerCase() in this.TEXTS) {
```
Comment (vladfrangu) [inline src/lib/Colors.ts:138]:
```suggestion
		if (text && Reflect.has(this.TEXTS, text.toLowerCase())) {
```

File: package.json [hunk 26]
```diff
@@ -50,5 +50,12 @@
...
       }
     }
-  }
+  },
+  "dependencies": {
+    "@klasa/timestamp": "^0.0.1",
+    "@klasa/utils": "^0.0.5"
+  },
+  "files": [
+    "dist/src/"
```
Comment (vladfrangu) [inline package.json:59]:
```suggestion
    "dist/src/"
```

File: src/lib/util/constants.ts [hunk 27]
```diff
@@ -0,0 +1,32 @@
+import { mergeDefault } from '@klasa/utils';
+import { ConsoleOptions, ConsoleOptionsColor } from '../KlasaConsole';
```
Comment (vladfrangu) [inline src/lib/util/constants.ts:2]:
```suggestion
import { mergeDefault } from '@klasa/utils';

import type { ConsoleOptions, ConsoleOptionsColor } from '../KlasaConsole';
```

File: package.json [hunk 28]
```diff
@@ -50,5 +50,12 @@
         "test/": "dist/test/"
       }
     }
-  }
+  },
+  "dependencies": {
```
Comment (MrJacz) [inline package.json:54]: Shouldn't these be peer deps then?

Comment (bdistin) [reply package.json:54]: I don't think so

Comment (MrJacz) [reply package.json:54]: According to vlad, they should be

**Pull request merged**